vizro-core/src/vizro/models/_action/_action.py

```diff
@@ -59,6 +69,50 @@ def build(self):
             "action_finished": Output("action_finished", "data", allow_duplicate=True),
         }
 
+        action_components = _get_action_callback_mapping(
+            action_id=self.id, argument="components"  # type: ignore[arg-type]
+        )
+
+        return callback_inputs, callback_outputs, action_components
+
+    def _action_callback_function(self, **inputs: Dict[str, Any]) -> Dict[str, Any]:
+        logger.debug("=============== ACTION ===============")
+        logger.debug(f'Action ID: "{self.id}"')
+        logger.debug(f'Action name: "{self.function._function.__name__}"')
+        logger.debug(f"Action inputs: {inputs}")
+
+        # Invoking the action's function
+        return_value = self.function(**inputs) or {}
+
+        # Raising the custom exception if return value length doesn't match the number of outputs
+        return_value_len = (
+            1 if not hasattr(return_value, "__len__") or isinstance(return_value, str) else len(return_value)
+        )
+        outputs = list(ctx.outputs_grouping.keys())
+        outputs.remove("action_finished")
+        if len(outputs) != return_value_len:
+            raise ValueError(
+                f"Number of action's returned elements: {return_value_len} does not match the number"
+                f" of action's defined outputs: {len(outputs)}."
+            )
+
+        # If return_value is a single element, ensure return_value is a list
+        if not isinstance(return_value, (list, tuple, dict)):
+            return_value = [return_value]
+        if isinstance(return_value, dict):
+            return {"action_finished": None, **return_value}
+
+        return {"action_finished": None, **dict(zip(outputs, return_value))}
+
+    @_log_call
+    def build(self):
+        """Builds a callback for the Action model and returns required components for the callback.
+
+        Returns:
+            List of required components (e.g. dcc.Download) for the Action model added to the `Dashboard` container.
```

Comment:
Is it possible/desirable to wrap these into a single `html.Div` or something else container? Until now we've had a policy that `build` should always return a single component and not a list of components. I know `Action` is a bit different, but it would be nice if we could maintain this convention if it's easy to do so.

Reply:
I managed to do it and now it looks much more consistent 😄. Also, the ActionLoop components are now bounded by an `html.Div` object and we can say that every application component (even every dcc.Store) has a parent Div.